File: plugin-sdk-ts/src/pkg.ts
import { callHost, callHostJson } from './host';
import type {
  Dependency,
  InstallReceipt,
  InstalledPackage,
  PackageInfo,
  PackageSummary,
  UpdateInfo,
  VersionInfo,
} from './types';

/**
 * Search for packages.
 * Requires: pkg_search permission.
 */
export function search(
  query: string,
  provider?: string | null,
): PackageSummary[] {
  return callHostJson<PackageSummary[]>(
    'cognia_pkg_search',
    JSON.stringify({ query, provider: provider ?? null }),
  );
}

/**
 * Get detailed package info.
 * Requires: pkg_search permission.
 */
export function info(
  name: string,
  provider?: string | null,
): PackageInfo {
  return callHostJson<PackageInfo>(
    'cognia_pkg_info',
    JSON.stringify({ name, provider: provider ?? null }),
  );
}

/**
 * Get available versions for a package.
 * Requires: pkg_search permission.
 */
export function versions(
  name: string,
  provider?: string | null,
): VersionInfo[] {
  return callHostJson<VersionInfo[]>(
    'cognia_pkg_versions',
    JSON.stringify({ name, provider: provider ?? null }),
  );
}

/**
 * Get dependencies for a specific package version.
 * Requires: pkg_search permission.
 */
export function dependencies(
  name: string,
  version: string,
  provider?: string | null,
): Dependency[] {
  return callHostJson<Dependency[]>(
    'cognia_pkg_dependencies',
    JSON.stringify({ name, version, provider: provider ?? null }),
  );
}

/**
 * List installed packages.
 * Requires: pkg_search permission.
 */
export function listInstalled(
  provider?: string | null,
): InstalledPackage[] {
  return callHostJson<InstalledPackage[]>(
    'cognia_pkg_list_installed',
    JSON.stringify({ provider: provider ?? null }),
  );
}

/**
 * Check for package updates.
 * Requires: pkg_search permission.
 */
export function checkUpdates(
  packages: string[],
  provider: string,
): UpdateInfo[] {
  return callHostJson<UpdateInfo[]>(
    'cognia_pkg_check_updates',
    JSON.stringify({ packages, provider }),
  );
}

/**
 * Install a package.
 * Requires: pkg_install permission.
 */
export function install(
  name: string,
  version?: string | null,
  provider?: string | null,
): InstallReceipt {
  return callHostJson<InstallReceipt>(
    'cognia_pkg_install',
    JSON.stringify({
      name,
      version: version ?? null,
      provider: provider ?? null,
    }),
  );
}

/**
 * Uninstall a package.
 * Requires: pkg_install permission.
 */
export function uninstall(
  name: string,
  version?: string | null,
  provider?: string | null,
): void {
  callHost(
    'cognia_pkg_uninstall',
    JSON.stringify({
      name,
      version: version ?? null,
      provider: provider ?? null,
    }),
  );
}

File: plugin-sdk/src/pkg.rs
use crate::host;
use crate::types::*;
use extism_pdk::*;

/// Search for packages.
pub fn search(query: &str, provider: Option<&str>) -> Result<Vec<PackageSummary>, Error> {
    let input = serde_json::json!({ "query": query, "provider": provider }).to_string();
    let result = unsafe { host::cognia_pkg_search(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get detailed package info.
pub fn info(name: &str, provider: Option<&str>) -> Result<PackageInfo, Error> {
    let input = serde_json::json!({ "name": name, "provider": provider }).to_string();
    let result = unsafe { host::cognia_pkg_info(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get available versions for a package.
pub fn versions(name: &str, provider: Option<&str>) -> Result<Vec<VersionInfo>, Error> {
    let input = serde_json::json!({ "name": name, "provider": provider }).to_string();
    let result = unsafe { host::cognia_pkg_versions(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Get dependencies for a specific package version.
pub fn dependencies(name: &str, version: &str, provider: Option<&str>) -> Result<Vec<Dependency>, Error> {
    let input = serde_json::json!({ "name": name, "version": version, "provider": provider }).to_string();
    let result = unsafe { host::cognia_pkg_dependencies(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// List installed packages.
pub fn list_installed(provider: Option<&str>) -> Result<Vec<InstalledPackage>, Error> {
    let input = serde_json::json!({ "provider": provider }).to_string();
    let result = unsafe { host::cognia_pkg_list_installed(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Check for package updates.
pub fn check_updates(packages: &[&str], provider: &str) -> Result<Vec<UpdateInfo>, Error> {
    let input = serde_json::json!({ "packages": packages, "provider": provider }).to_string();
    let result = unsafe { host::cognia_pkg_check_updates(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Install a package.
pub fn install(name: &str, version: Option<&str>, provider: Option<&str>) -> Result<InstallReceipt, Error> {
    let input = serde_json::json!({ "name": name, "version": version, "provider": provider }).to_string();
    let result = unsafe { host::cognia_pkg_install(input)? };
    Ok(serde_json::from_str(&result)?)
}

/// Uninstall a package.
pub fn uninstall(name: &str, version: Option<&str>, provider: Option<&str>) -> Result<(), Error> {
    let input = serde_json::json!({ "name": name, "version": version, "provider": provider }).to_string();
    unsafe { host::cognia_pkg_uninstall(input)?; }
    Ok(())
}
